In [ ]:
import pandas as pd
import plotly.graph_objects as go
import numpy as np
import holidays

In [5]:
X_data_df = pd.read_csv("../Datos/X.csv", sep=",")
Y_data_df = pd.read_csv("../Datos/Y.csv", sep=";")
Z_data_df = pd.read_csv("../Datos/Z.csv", sep=",")

## Explorar formatos

#### Formato "date"

In [6]:
X_data_df['Date'] = pd.to_datetime(X_data_df['Date'], format="%Y-%m-%d")
Y_data_df['Date'] = pd.to_datetime(Y_data_df['Date'], format="%d/%m/%Y")
Z_data_df['Date'] = pd.to_datetime(Z_data_df['Date'], format="%Y-%m-%d")

#### Formato "float"

In [7]:
if str(Y_data_df['Price'].dtype) == "object":
    Y_data_df['Price'] = (Y_data_df['Price'].str.replace(",", ".")).astype(float)
if str(X_data_df['Price'].dtype) == "object":
    X_data_df['Price'] = (X_data_df['Price'].str.replace(",", ".")).astype(float)
if str(Z_data_df['Price'].dtype) == "object":
    Z_data_df['Price'] = (Z_data_df['Price'].str.replace(",", ".")).astype(float)

#### Reordenar

In [8]:
X_data_df = X_data_df.sort_values(by="Date")
X_data_df = X_data_df.reset_index(drop=True)
Y_data_df = Y_data_df.sort_values(by="Date")
Y_data_df = Y_data_df.reset_index(drop=True)
Z_data_df = Z_data_df.sort_values(by="Date")
Z_data_df = Z_data_df.reset_index(drop=True)

## Explorar datos

In [9]:
print("X_data_df:", X_data_df["Date"].min(), X_data_df["Date"].max())
print("Y_data_df:", Y_data_df["Date"].min(), Y_data_df["Date"].max())
print("Z_data_df:", Z_data_df["Date"].min(), Z_data_df["Date"].max())

X_data_df: 1988-06-27 00:00:00 2024-04-04 00:00:00
Y_data_df: 2006-07-11 00:00:00 2023-09-12 00:00:00
Z_data_df: 2010-01-01 00:00:00 2023-08-31 00:00:00


In [10]:
print("X_data_df Fechas duplicadas:", X_data_df.duplicated(subset=["Date"]).sum())
print("X_data_df Fechas duplicadas:", Y_data_df.duplicated(subset=["Date"]).sum())
print("X_data_df Fechas duplicadas:", Z_data_df.duplicated(subset=["Date"]).sum())

X_data_df Fechas duplicadas: 0
X_data_df Fechas duplicadas: 0
X_data_df Fechas duplicadas: 0


In [11]:
print("X_data_df nulos:", X_data_df.isnull().sum())
print("Y_data_df nulos:", Y_data_df.isnull().sum())
print("Z_data_df nulos:", Z_data_df.isnull().sum())

X_data_df nulos: Date     0
Price    0
dtype: int64
Y_data_df nulos: Date     0
Price    0
dtype: int64
Z_data_df nulos: Price    0
Date     0
dtype: int64


#### Frecuencia temporal

In [12]:
# X
X_data_df["diff_days"] = X_data_df["Date"].diff().dt.days
X_most_common_diff = X_data_df["diff_days"].mode()
X_data_df["diff_days"] = X_data_df["Date"].diff().dt.days
X_most_common_diff = X_data_df["diff_days"].mode()
# Y
Y_data_df["diff_days"] = Y_data_df["Date"].diff().dt.days
Y_most_common_diff = Y_data_df["diff_days"].mode()
Y_data_df["diff_days"] = Y_data_df["Date"].diff().dt.days
Y_most_common_diff = Y_data_df["diff_days"].mode()
# Z
Z_data_df["diff_days"] = Z_data_df["Date"].diff().dt.days
Z_most_common_diff = Z_data_df["diff_days"].mode()
Z_data_df["diff_days"] = Z_data_df["Date"].diff().dt.days
Z_most_common_diff = Z_data_df["diff_days"].mode()

In [13]:
print("X_data_df Días entre registros (frecuencia más común):", X_most_common_diff.values if not X_most_common_diff.empty else "No determinada")
print("Y_data_df Días entre registros (frecuencia más común):", Y_most_common_diff.values if not Y_most_common_diff.empty else "No determinada")
print("Z_data_df Días entre registros (frecuencia más común):", Z_most_common_diff.values if not Z_most_common_diff.empty else "No determinada")

X_data_df Días entre registros (frecuencia más común): [1.]
Y_data_df Días entre registros (frecuencia más común): [1.]
Z_data_df Días entre registros (frecuencia más común): [1.]


#### Fechas faltantes

In [14]:
# X
X_full_range = pd.date_range(start=X_data_df["Date"].min(), end=X_data_df["Date"].max(), freq="D")
X_missing_dates = X_full_range.difference(X_data_df["Date"])
print("X_data_df Fechas faltantes (frecuencia diaria esperada):", len(X_missing_dates))
# Y
Z_full_range = pd.date_range(start=Z_data_df["Date"].min(), end=Z_data_df["Date"].max(), freq="D")
Z_missing_dates = Z_full_range.difference(Z_data_df["Date"])
print("Z_data_df Fechas faltantes (frecuencia diaria esperada):", len(Z_missing_dates))
# Z
Z_full_range = pd.date_range(start=Z_data_df["Date"].min(), end=Z_data_df["Date"].max(), freq="D")
Z_missing_dates = Z_full_range.difference(Z_data_df["Date"])
print("Z_data_df Fechas faltantes (frecuencia diaria esperada):", len(Z_missing_dates))

X_data_df Fechas faltantes (frecuencia diaria esperada): 3922
Z_data_df Fechas faltantes (frecuencia diaria esperada): 1426
Z_data_df Fechas faltantes (frecuencia diaria esperada): 1426


#### Volatilidad (desviación estándar)

In [16]:
# X
X_data_df["return"] = X_data_df["Price"].pct_change()
X_volatility = X_data_df["return"].std()
print("X_data_df Volatilidad (desviación estándar):", X_volatility)
# Y
Y_data_df["return"] = Y_data_df["Price"].pct_change()
Y_volatility = Y_data_df["return"].std()
print("Y_data_df Volatilidad (desviación estándar):", Y_volatility)
# Z
Z_data_df["return"] = Z_data_df["Price"].pct_change()
Z_volatility = Z_data_df["return"].std()
print("Z_data_df Volatilidad (desviación estándar):", Z_volatility)

X_data_df Volatilidad (desviación estándar): 0.022818459068399792
Y_data_df Volatilidad (desviación estándar): 0.012559041023127708
Z_data_df Volatilidad (desviación estándar): 0.013674384846492128


#### Conteo por día de la semana 

Debido a la frecuencia diaria esperada que encontramos

In [17]:
X_data_df["weekday"] = X_data_df["Date"].dt.day_name()
X_conteo_dias = X_data_df["weekday"].value_counts()
print("X_data_df:\n", X_conteo_dias)
Y_data_df["weekday"] = Y_data_df["Date"].dt.day_name()
Y_conteo_dias = Y_data_df["weekday"].value_counts()
print("Y_data_df:\n", Y_conteo_dias)
Z_data_df["weekday"] = Z_data_df["Date"].dt.day_name()
Z_conteo_dias = Z_data_df["weekday"].value_counts()
print("Z_data_df:\n", Z_conteo_dias)

X_data_df:
 weekday
Thursday     1856
Wednesday    1856
Tuesday      1848
Friday       1816
Monday       1766
Sunday          2
Name: count, dtype: int64
Y_data_df:
 weekday
Tuesday      897
Wednesday    896
Thursday     896
Friday       896
Monday       896
Saturday       4
Name: count, dtype: int64
Z_data_df:
 weekday
Friday       713
Monday       713
Tuesday      713
Wednesday    713
Thursday     713
Name: count, dtype: int64


#### Revisar festivos

Con los anteriores datos se puede intuir que solo se emplea el uso de las materias primas en días hábiles, para confirmar aún más la inferencia, revisaremos en cuántos festivos se han han empleado materias primas.

In [18]:
min_X_data_df = X_data_df['Date'].min()
max_X_data_df = X_data_df['Date'].max()
min_Y_data_df = Y_data_df['Date'].min()
max_Y_data_df = Y_data_df['Date'].max()
min_Z_data_df = Z_data_df['Date'].min()
max_Z_data_df = Z_data_df['Date'].max()

In [19]:
start_year_X_data = int(min_X_data_df.strftime("%Y"))
end_year_X_data = int(max_X_data_df.strftime("%Y"))
co_holidays = holidays.Colombia(years=range(start_year_X_data, end_year_X_data + 1))

holidays_df = pd.DataFrame(
    [(date, name) for date, name in co_holidays.items()],
    columns=["Date", "Holiday"]
)
holidays_df["Date"] = pd.to_datetime(holidays_df["Date"])
holidays_df = holidays_df[holidays_df["Date"].dt.weekday != 6]

In [20]:
def get_holidays_data(validate_df, nombre_df):
    validate_df = validate_df.copy()
    validate_df["Date"] = pd.to_datetime(validate_df["Date"])
    
    validate_df = validate_df.merge(
        holidays_df,
        on="Date",
        how="left"
    )
    
    validate_df["es_festivo"] = validate_df["Holiday"].notna()
    
    print(f"\nResumen {nombre_df}:")
    print(validate_df["es_festivo"].value_counts())
    
    return validate_df

X_data_df = get_holidays_data(X_data_df, "X")
Y_data_df = get_holidays_data(Y_data_df, "Y")
Z_data_df = get_holidays_data(Z_data_df, "Z")


Resumen X:
es_festivo
False    8663
True      481
Name: count, dtype: int64

Resumen Y:
es_festivo
False    4208
True      277
Name: count, dtype: int64

Resumen Z:
es_festivo
False    3348
True      217
Name: count, dtype: int64


In [21]:
print("Min: ", min_X_data_df, min_Y_data_df, min_Z_data_df)
print("Max: ", max_X_data_df, max_Y_data_df, max_Z_data_df)

Min:  1988-06-27 00:00:00 2006-07-11 00:00:00 2010-01-01 00:00:00
Max:  2024-04-04 00:00:00 2023-09-12 00:00:00 2023-08-31 00:00:00


In [22]:
X_data_df[X_data_df["es_festivo"]]["Holiday"].value_counts()

Holiday
Jueves Santo                                                      36
Corpus Christi (observado)                                        32
Día de la Raza (observado)                                        31
Día de San José (observado)                                       31
Independencia de Cartagena (observado)                            31
Día de los Reyes Magos (observado)                                31
Día de Todos los Santos (observado)                               31
Ascensión del señor (observado)                                   30
La Asunción (observado)                                           30
Sagrado Corazón (observado)                                       29
Día de la Independencia                                           27
San Pedro y San Pablo (observado)                                 26
La Inmaculada Concepción                                          26
Batalla de Boyacá                                                 25
Día del Trabajo           

Con esto nos damos cuenta de que se ha trabajado en festivos colombianos.

#### Visualización de datos en gráfico de líneas

In [23]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=X_data_df["Date"],
    y=X_data_df["Price"],
    mode='lines',
    name='X'
))

fig.add_trace(go.Scatter(
    x=Y_data_df["Date"],
    y=Y_data_df["Price"],
    mode='lines',
    name='Y'
))

fig.add_trace(go.Scatter(
    x=Z_data_df["Date"],
    y=Z_data_df["Price"],
    mode='lines',
    name='Z'
))

fig.update_layout(
    title="Comparación series temporales",
    xaxis_title="Fecha",
    yaxis_title="Precio",
    template="plotly_white",
    width=1100,
    height=600
)

fig.update_xaxes(
    tickformatstops=[
        dict(dtickrange=[None, "M1"], value="%d %b %Y"),
        dict(dtickrange=["M1", "M12"], value="%b %Y"),
        dict(dtickrange=["M12", None], value="%Y")
    ]
)

fig.show()

## Modelos

Ahora vamos a adaptar los datos a como lo manejaremos y empezaremos a entrenar cada modelo por materia prima.

De acuerdo a lo que vimos, vamos a tratar dos secciones:
1. Usar log-retornos (aplicación logarítmica a los precios) para explorar su comportamiento.
2. Usar precios para comparar frente a log-retornos si ajusta mejor.

In [24]:
import warnings
warnings.filterwarnings("ignore")

from statsmodels.tsa.holtwinters import ExponentialSmoothing
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.tsa.stattools import adfuller

from sklearn.metrics import mean_absolute_error, mean_squared_error

In [25]:
X_data_process_df = X_data_df[["Date", "Price"]].copy()
Y_data_process_df = Y_data_df[["Date", "Price"]].copy()
Z_data_process_df = Z_data_df[["Date", "Price"]].copy()

In [26]:
def prepare_serie(data_df):
    data_df = data_df.copy()
    data_df["Date"] = pd.to_datetime(data_df["Date"])
    data_df = data_df.sort_values("Date")
    data_df = data_df.set_index("Date")
    data_df = data_df.asfreq("B")
    data_df["Price"] = data_df["Price"].ffill()
    return data_df

X_data_process_df = prepare_serie(X_data_process_df)
Y_data_process_df = prepare_serie(Y_data_process_df)
Z_data_process_df = prepare_serie(Z_data_process_df)

In [27]:
print(X_data_process_df.shape)
print(Y_data_process_df.shape)
print(Z_data_process_df.shape)

(9334, 1)
(4481, 1)
(3565, 1)


### Sección 1. Modelado con log-retornos

In [24]:
def calculate_log_return(data_df):
    data_df = data_df.copy()
    data_df["log_return"] = np.log(data_df["Price"] / data_df["Price"].shift(1))
    return data_df.dropna()

X_data_process_df = calculate_log_return(X_data_process_df)
Y_data_process_df = calculate_log_return(Y_data_process_df)
Z_data_process_df = calculate_log_return(Z_data_process_df)

print(X_data_process_df.shape)
print(Y_data_process_df.shape)
print(Z_data_process_df.shape)

(9333, 2)
(4480, 2)
(3564, 2)


In [25]:
X_result = adfuller(X_data_process_df["log_return"])
print("p-value:", X_result[1])
Y_result = adfuller(Y_data_process_df["log_return"])
print("p-value:", Y_result[1])
Z_result = adfuller(Z_data_process_df["log_return"])
print("p-value:", Z_result[1])

p-value: 0.0
p-value: 5.195705863477862e-17
p-value: 7.747030892777954e-30


Esto lo realziamos para revisar si es estacionario, pues p-value < 0.05 nos habla de que la serie es estacionaria. Encontramos que los Log-retornos (aplicado logarítimicamente a los precios) son estacioarios.

#### Modelos preliminares

Se pronosticará 1 año para evaluar el desempeño del modelo; al tratar únicamente días hábiles, se proyectarán 252 observaciones. A los precios históricos se les aplicó transformación de log-retorno con el fin de estabilizar la varianza y garantizar estacionariedad, validada mediante la prueba ADF (Augmented Dickey-Fuller) anteriormente realizara con las materias primas y el cálculo de p-value.

Para ser eficientes en tiempo, vamos a realizar dos momentos en la evaluación de los modelos:
1. Se realizara una evaluación con el modelo ARIMA con múltiples parámetros de acuerdo a la métrica AIC que es recomendada para más rápida validación de mejores modelos e hiperparámetros, frente a otras métricas.
2. Con los mejores 3 hiperparámetros, entrenaremos el modelo ARIMA (comúnmente mejor que ETS para estos casos) y pasaremos a pronosticar con las 252 observaciones.
   
    Seleccionaremos el mejor y continuaremos en la sección del pronostico de los 36 meses.

_No trataremos modelos como SARIMAX debido a que los log-retornos son estacionarios (ADF p-value < 0.05) y no hay una demostración de estacionalidad significativa._

Los modelos se compararán utilizando:
1. MAE: Error absoluto promedio.
2. RMSE: Penaliza con mayor peso los errores grandes.

#### Funciones

In [29]:
def select_by_aic(serie_log, p_values, q_values):

    serie_log = serie_log.dropna().reset_index(drop=True)
    results = []

    for p in p_values:
        for q in q_values:
            try:
                model = ARIMA(serie_log, order=(p,0,q))
                result = model.fit(method="innovations_mle")
                results.append((p,q,result.aic))
            except:
                continue

    data_df = pd.DataFrame(results, columns=["p","q","AIC"])
    return data_df.sort_values("AIC").reset_index(drop=True)

In [ ]:
def rolling_cv_top_models(
    serie_log, model_list, initial_train_size=0.6, forecast_horizon=21, step=21):
    serie_log = serie_log.dropna().reset_index(drop=True)
    n = len(serie_log)
    train_size = int(n * initial_train_size)
    results = []

    for (p,q) in model_list:
        maes = []
        rmses = []
        start = train_size
        print(p,q)

        while start + forecast_horizon <= n:
            train = serie_log.iloc[:start]
            test = serie_log.iloc[start:start+forecast_horizon]
            try:
                model = ARIMA(train, order=(p,0,q))
                result = model.fit(method="innovations_mle")
                forecast = result.forecast(steps=forecast_horizon)

                mae = mean_absolute_error(test, forecast)
                rmse = np.sqrt(mean_squared_error(test, forecast))
                maes.append(mae)
                rmses.append(rmse)
            except:
                break
            start += step
        if maes:
            results.append({
                "p": p,
                "q": q,
                "MAE": np.mean(maes),
                "RMSE": np.mean(rmses)
            })

    return pd.DataFrame(results).sort_values("RMSE").reset_index(drop=True)

#### Ejecución para pronosticar 1 año de las materias primas

##### Materia prima X

In [ ]:
p_values = [0,1,2,3]
q_values = [0,1,2,3]
serie_X_log = X_data_process_df["log_return"]

X_aic_df = select_by_aic(serie_X_log, p_values, q_values)
X_top_models = X_aic_df.head(3)[["p","q"]].values.tolist()
print(X_top_models)

[[2, 3], [3, 2], [3, 3]]


In [35]:
tabla_cv_X = rolling_cv_top_models(
    serie_X_log,
    X_top_models
)
print(tabla_cv_X)

   p  q       MAE      RMSE
0  3  2  0.015042  0.019467
1  2  3  0.015044  0.019470
2  3  3  0.015049  0.019475


In [43]:
best_q_X = tabla_cv_X["q"].values[0]
best_p_X = tabla_cv_X["p"].values[0]

##### Materia prima Y 

In [36]:
p_values = [0,1,2,3]
q_values = [0,1,2,3]
serie_Y_log = Y_data_process_df["log_return"]

Y_aic_df = select_by_aic(serie_Y_log, p_values, q_values)
Y_top_models = Y_aic_df.head(3)[["p","q"]].values.tolist()
print(Y_top_models)

[[3, 3], [2, 3], [3, 2]]


In [37]:
tabla_cv_Y = rolling_cv_top_models(
    serie_Y_log,
    Y_top_models
)
print(tabla_cv_Y)

   p  q       MAE      RMSE
0  2  3  0.004354  0.007291
1  3  3  0.004542  0.007539
2  3  2  0.008802  0.020657


In [38]:
best_q_Y = tabla_cv_Y["q"].values[0]
best_p_Y = tabla_cv_Y["p"].values[0]

##### Materia prima Z

In [39]:
p_values = [0,1,2,3]
q_values = [0,1,2,3]
serie_Z_log = Z_data_process_df["log_return"]

Z_aic_df = select_by_aic(serie_Z_log, p_values, q_values)
Z_top_models = Z_aic_df.head(3)[["p","q"]].values.tolist()
print(Z_top_models)

[[2, 2], [3, 3], [3, 2]]


In [40]:
tabla_cv_Z = rolling_cv_top_models(
    serie_Z_log,
    Z_top_models
)
print(tabla_cv_Z)

   p  q       MAE      RMSE
0  2  2  0.010804  0.013832
1  3  2  0.010829  0.013844
2  3  3  0.010841  0.013864


In [ ]:
best_q_Z = tabla_cv_Z["q"].values[0]
best_p_Z = tabla_cv_Z["p"].values[0]
print("Mejores modelos:", best_p_X, best_q_X, best_p_Y, best_q_Y, best_p_Z, best_q_Z)

##### Conclusiones preliminares

In [46]:
print("Mejores modelos (p, q):", "X:", best_p_X, best_q_X, "Y:", best_p_Y, best_q_Y, "Z:", best_p_Z, best_q_Z)

Mejores modelos (p, q): X: 3 2 Y: 2 3 Z: 2 2


Con lo anterior, encontramos buenos resultados de los errores para cada modelo, con esto vamos a llevar los datos a los modelos que pronosticarán lo que requerimos.

#### Proceso completo para entrenamiento de los modelos 

Vamos a explorar con log-retornos el enfoque que hemos venido teniendo.

##### Calcular días hábiles exactos a pronosticar (36 meses)

In [102]:
start_date = X_data_process_df.index.max()
end_date = start_date + pd.DateOffset(months=36)

business_days = pd.bdate_range(start=start_date + pd.Timedelta(days=1), end=end_date)
steps_total = len(business_days)

print("Fecha inicio:", start_date)
print("Fecha fin:", end_date)
print("Días hábiles a pronosticar:", steps_total)

Fecha inicio: 2024-04-04 00:00:00
Fecha fin: 2027-04-04 00:00:00
Días hábiles a pronosticar: 781


##### Llevar a las materias primas Y y Z a la misma fecha de inicio del proyecto (basándonos en la fecha fin de registros de la materia prima X)

In [103]:
print("Última fecha X:", max_X_data_df)
print("Última fecha Y:", max_Y_data_df)
print("Última fecha Z:", max_Z_data_df)

gap_Y = int(np.busday_count(max_Y_data_df.date(), max_X_data_df.date()))
gap_Z = int(np.busday_count(max_Z_data_df.date(), max_X_data_df.date()))

print("Días hábiles que Y debe proyectar:", gap_Y)
print("Días hábiles que Z debe proyectar:", gap_Z)

Última fecha X: 2024-04-04 00:00:00
Última fecha Y: 2023-09-12 00:00:00
Última fecha Z: 2023-08-31 00:00:00
Días hábiles que Y debe proyectar: 147
Días hábiles que Z debe proyectar: 155


Alineación materia prima Y

In [62]:
model_Y = ARIMA(serie_Y_log, order=(best_p_Y,0,best_q_Y)).fit(method="innovations_mle")
forecast_Y_gap = model_Y.forecast(steps=gap_Y)

serie_Y_extended = pd.concat([serie_Y_log, forecast_Y_gap])

Alineación materia prima Z

In [63]:
model_Z = ARIMA(serie_Z_log, order=(best_p_Z,0,best_q_Z)).fit(method="innovations_mle")
forecast_Z_gap = model_Z.forecast(steps=gap_Z)

serie_Z_extended = pd.concat([serie_Z_log, forecast_Z_gap])

##### Pronosticar 36 meses

In [104]:
best_model_X = ARIMA(serie_X_log, order=(best_p_X,0,best_q_X)).fit(method="innovations_mle")
forecast_X_log = best_model_X.forecast(steps=781)

In [105]:
model_Y_updated = ARIMA(serie_Y_extended, order=(best_p_Y,0,best_q_Y)).fit(method="innovations_mle")
forecast_Y_log = model_Y_updated.forecast(steps=781)

In [106]:
model_Z_updated = ARIMA(serie_Z_extended, order=(best_p_Z,0,best_q_Z)).fit(method="innovations_mle")
forecast_Z_log = model_Z_updated.forecast(steps=781)

##### Reconstrucción de los precios desde log-retornos

Tal como calculamos los log-retornos a partir de los precios, ahora realizaremos lo inverso para tener de nuevo los precios.
Fórmula:
`Pt​=Pt−1​e^(rt)​`

In [84]:
def reconstruct_prices(last_price, forecast_log):
    prices = [last_price]
    for r in forecast_log:
        next_price = prices[-1] * np.exp(r)
        prices.append(next_price)
    return prices[1:]
# X
forecast_X_prices = reconstruct_prices(X_data_process_df["Price"].values[-1], forecast_X_log)
# Y
last_real_price_Y = Y_data_process_df["Price"].iloc[-1]
gap_prices_Y = reconstruct_prices(
    last_real_price_Y,
    forecast_Y_gap
)
last_aligned_price_Y = gap_prices_Y[-1]
forecast_Y_prices = reconstruct_prices(last_aligned_price_Y, forecast_Y_log)
# Z
last_real_price_Z = Z_data_process_df["Price"].iloc[-1]
gap_prices_Z = reconstruct_prices(
    last_real_price_Z,
    forecast_Z_gap
)
last_aligned_price_Z = gap_prices_Z[-1]
forecast_Z_prices = reconstruct_prices(last_aligned_price_Z, forecast_Z_log)

##### Visualicemos cómo nos quedaría

Se procesarán dos meses adicionales para ver comportamiento (826 días habiles)

In [ ]:
future_dates = pd.bdate_range(
    start=X_data_process_df.index.max() + pd.Timedelta(days=1),
    periods=826
)

In [ ]:
forecast_X_df = pd.DataFrame({
    "Date": future_dates,
    "Price": forecast_X_prices
}).set_index("Date")

In [85]:
gap_dates_Y = pd.bdate_range(
    start=Y_data_process_df.index.max() + pd.Timedelta(days=1),
    periods=gap_Y
)
gap_Y_df = pd.DataFrame({
    "Date": gap_dates_Y,
    "Price": gap_prices_Y
}).set_index("Date")
forecast_Y_df = pd.DataFrame({
    "Date": future_dates,
    "Price": forecast_Y_prices
}).set_index("Date")

In [86]:
gap_dates_Z = pd.bdate_range(
    start=Z_data_process_df.index.max() + pd.Timedelta(days=1),
    periods=gap_Z
)
gap_Z_df = pd.DataFrame({
    "Date": gap_dates_Z,
    "Price": gap_prices_Z
}).set_index("Date")
forecast_Z_df = pd.DataFrame({
    "Date": future_dates,
    "Price": forecast_Z_prices
}).set_index("Date")

In [91]:
def plot_real_gap_forecast(real_df, forecast_df, title, gap_df=None):
    fig = go.Figure()

    fig.add_trace(go.Scatter( # Histórico
        x=real_df.index,
        y=real_df["Price"],
        mode="lines",
        name="Histórico"
    ))
    if gap_df is not None:
        fig.add_trace(go.Scatter( # GAP puente
            x=gap_df.index,
            y=gap_df["Price"],
            mode="lines",
            name="Alineación temporal",
            line=dict()
        ))
    fig.add_trace(go.Scatter( # Forecast oficial
        x=forecast_df.index,
        y=forecast_df["Price"],
        mode="lines",
        name="Pronóstico",
        line=dict()
    ))

    fig.update_layout(
        title=title,
        template="plotly_white"
    )
    fig.show()

In [92]:
plot_real_gap_forecast(X_data_process_df, forecast_X_df, "Materia Prima X")
plot_real_gap_forecast(Y_data_process_df, forecast_Y_df, "Materia Prima Y", gap_Y_df)
plot_real_gap_forecast(Z_data_process_df, forecast_Z_df, "Materia Prima Z", gap_Z_df)

##### Construcción de equipos

- Equipo 1: `0.2X + 0.8Y`
- Equipo 2: `(X + Y + Z) / 3`

In [95]:
team_1 = 0.2*np.array(forecast_X_prices) + 0.8*np.array(forecast_Y_prices)
team_2 = (np.array(forecast_X_prices) + np.array(forecast_Y_prices) + np.array(forecast_Z_prices)) / 3
print("Costo preliminar del equipo 1:", np.mean(team_1))
print("Costo preliminar del equipo 2:", np.mean(team_2))

Costo preliminar del equipo 1: 451.75833445865766
Costo preliminar del equipo 2: 930.4706988803423


##### Aplicación de la segunda parte (Análisis de escenarios)

Para esto aplicaremos la fórmula de Monte Carlo

In [96]:
def monte_carlo_arima_pro(
    model, last_price, steps, simulations=1000):
    simulated_price_paths = []

    for i in range(simulations):
        simulated_returns = model.simulate(
            nsimulations=steps
        )
        prices = [last_price]

        for r in simulated_returns:
            prices.append(prices[-1] * np.exp(r))

        simulated_price_paths.append(prices[1:])

    return np.array(simulated_price_paths)

monte_carlo_sim_X = monte_carlo_arima_pro(
    best_model_X, X_data_process_df["Price"].iloc[-1], steps=826, simulations=1000
)

monte_carlo_sim_Y = monte_carlo_arima_pro(
    model_Y_updated, last_aligned_price_Y, steps=826, simulations=1000
)

monte_carlo_sim_Z = monte_carlo_arima_pro(
    model_Z_updated, last_aligned_price_Z, steps=826, simulations=1000
)

In [107]:
monte_carlo_team_1 = 0.2*monte_carlo_sim_X + 0.8*monte_carlo_sim_Y
monte_carlo_team_2 = (monte_carlo_sim_X + monte_carlo_sim_Y + monte_carlo_sim_Z) / 3
print("Costo preliminar del equipo 1:", np.mean(monte_carlo_team_1))
print("Costo preliminar del equipo 2:", np.mean(monte_carlo_team_2))

Costo preliminar del equipo 1: 497.91449527020706
Costo preliminar del equipo 2: 982.2698016652531


##### Métricas sobre los costos

In [110]:
def get_metrics(monte_carlo_matrix):
    final_values = monte_carlo_matrix[:, -1]
    return {
        "Media": np.mean(final_values),
        "Desviación estandard": np.std(final_values),
        "Min": np.min(final_values),
        "Max": np.max(final_values),
        "P5": np.percentile(final_values, 5),
        "P50": np.percentile(final_values, 50),
        "P95": np.percentile(final_values, 95)
    }

metrics_team_1 = get_metrics(monte_carlo_team_1)
print(metrics_team_1)
metrics_team_2 = get_metrics(monte_carlo_team_2)
print(metrics_team_2)

{'Media': np.float64(542.7851065749583), 'Desviación estandard': np.float64(340.1965070691008), 'Min': np.float64(94.48811180636469), 'Max': np.float64(3207.0516036599693), 'P5': np.float64(183.7058283473736), 'P50': np.float64(463.0037409904253), 'P95': np.float64(1143.9712343221088)}
{'Media': np.float64(1030.6788744278806), 'Desviación estandard': np.float64(323.0256741132682), 'Min': np.float64(415.85587273700565), 'Max': np.float64(2666.39190356794), 'P5': np.float64(616.4640367740011), 'P50': np.float64(979.9173289215532), 'P95': np.float64(1639.4390947222403)}


In [111]:
def plot_monte_carlo(mean_path, p5, p95, dates, title):
    fig = go.Figure()
    fig.add_trace(go.Scatter(
        x=dates,
        y=p95,
        line=dict(width=0),
        showlegend=False
    ))
    fig.add_trace(go.Scatter(
        x=dates,
        y=p5,
        fill='tonexty',
        name="Banda 90%",
        line=dict(width=0)
    ))
    fig.add_trace(go.Scatter(
        x=dates,
        y=mean_path,
        name="Escenario esperado",
        line=dict(width=2)
    ))

    fig.update_layout(
        title=title,
        template="plotly_white"
    )
    fig.show()

plot_monte_carlo(
    monte_carlo_team_1.mean(axis=0),
    np.percentile(monte_carlo_team_1, 5, axis=0),
    np.percentile(monte_carlo_team_1, 95, axis=0),
    future_dates,
    "Equipo 1"
)

plot_monte_carlo(
    monte_carlo_team_2.mean(axis=0),
    np.percentile(monte_carlo_team_2, 5, axis=0),
    np.percentile(monte_carlo_team_2, 95, axis=0),
    future_dates,
    "Equipo 2"
)

Como podemos observar, con log-retornos, tenemos un escenario esperado incremental con una pendiente muy pequeña y una banda de confianza demasiado alta.

Vamos a explorar tratando directamente los precios a ver qué tal.

### Sección 2. Modelado con precios

Validación en niveles de precios. Con esto esperamos conseguir p-value > 0.05 : No estacionaria.

In [28]:
def adf_test(series, name):
    result = adfuller(series.dropna())
    print(f"{name} - p-value:", result[1])
def adf_test_ct(series, name):
    result = adfuller(series, regression="ct")
    print(f"{name} - p-value:", result[1])

adf_test(X_data_process_df["Price"], "X Price")
adf_test(Y_data_process_df["Price"], "Y Price")
adf_test_ct(Y_data_process_df["Price"], "Y Price ct")
adf_test(Z_data_process_df["Price"], "Z Price")

X Price - p-value: 0.2334605125815747
Y Price - p-value: 0.013406480163570567
Y Price ct - p-value: 0.060352384555825776
Z Price - p-value: 0.18410157066210997


Con esto encontramos que las materias primas X y Z no son estacionarias y la materia prima Y al parecer es estacionaria, pero si luego la tratamos con tendencia no se sostiene la hipótesis al 5%. Lo que nos lleva a considerar las materias primas como L(1), diferenciación de primer orden.

En este caso usar Arima con Drift es ideal frente a ETS o SARIMA, principalmente gracias a la evaluación del p-value y al enfoque que requiere este caso

#### Modelos preliminares

In [29]:
def select_arima_levels(series, p_values=[0,1,2], q_values=[0,1,2]):
    results = []

    for p in p_values:
        for q in q_values:
            try:
                model = ARIMA(series, order=(p,1,q), trend="t")
                fitted = model.fit(method="innovations_mle")
                results.append((p,q,fitted.aic))
            except:
                continue
    data_df = pd.DataFrame(results, columns=["p","q","AIC"])
    return data_df.sort_values("AIC").reset_index(drop=True)

In [30]:
X_aic_levels = select_arima_levels(X_data_process_df["Price"])
Y_aic_levels = select_arima_levels(Y_data_process_df["Price"])
Z_aic_levels = select_arima_levels(Z_data_process_df["Price"])

print(X_aic_levels.head())
print(Y_aic_levels.head())
print(Z_aic_levels.head())

   p  q           AIC
0  2  2  30277.873583
1  2  0  30283.262249
2  0  2  30283.359256
3  2  1  30284.232013
4  1  1  30284.347770
   p  q           AIC
0  1  1  30891.417637
1  2  2  30893.099969
2  1  2  30893.414999
3  2  1  30893.415089
4  2  0  30931.678698
   p  q           AIC
0  2  2  34561.084277
1  1  1  34578.299231
2  2  0  34578.336206
3  1  2  34578.658028
4  2  1  34578.862265


Con estos resultados, concluimos que:
1. Materia prima X: Usaremos p:2 y q:2.
2. Materia prima Y: Usaremos p:1 y q:1
3. Materia prima Z: Usaremos p:2 y q:2

Debido que no hay una diferencia no significativa que nos ponga a dudar a tomar varios p y q.

In [31]:
arima_X = ARIMA(
    X_data_process_df["Price"],
    order=(2,1,2),
    trend="t"
).fit(method="innovations_mle")

arima_Y = ARIMA(
    Y_data_process_df["Price"],
    order=(1,1,1),
    trend="t"
).fit(method="innovations_mle")

arima_Z = ARIMA(
    Z_data_process_df["Price"],
    order=(2,1,2),
    trend="t"
).fit(method="innovations_mle")

#### Ejecución para pronosticar 1 año de las materias primas

In [32]:
def evaluate_arima_levels(series, order, test_days=252):
    train = series[:-test_days]
    test = series[-test_days:]
    model = ARIMA(train, order=order, trend="t").fit(method="innovations_mle")
    forecast = model.forecast(steps=test_days)

    mae = mean_absolute_error(test, forecast)
    rmse = np.sqrt(mean_squared_error(test, forecast))
    mape = np.mean(np.abs((test - forecast) / test)) * 100

    return mae, rmse, mape

In [33]:
metrics_X = evaluate_arima_levels(
    X_data_process_df["Price"],
    order=(2,1,2)
)

metrics_Y = evaluate_arima_levels(
    Y_data_process_df["Price"],
    order=(1,1,1)
)

metrics_Z = evaluate_arima_levels(
    Z_data_process_df["Price"],
    order=(2,1,2)
)

print("Materia prima X - MAE, RMSE, MAPE:", metrics_X)
print("Materia prima Y - MAE, RMSE, MAPE:", metrics_Y)
print("Materia prima Z - MAE, RMSE, MAPE:", metrics_Z)

Materia prima X - MAE, RMSE, MAPE: (5.567790500521356, np.float64(6.543380892825871), np.float64(6.99549093940731))
Materia prima Y - MAE, RMSE, MAPE: (34.33614586822283, np.float64(45.727341829973525), np.float64(5.670918285172987))
Materia prima Z - MAE, RMSE, MAPE: (102.49687181057944, np.float64(126.52646961336285), np.float64(4.5046728115309715))


In [34]:
def plot_evaluation(series, order, test_days=252, title=""):
    train = series[:-test_days]
    test = series[-test_days:]

    model = ARIMA(train, order=order, trend="t").fit(method="innovations_mle")
    forecast = model.forecast(steps=test_days)
    fig = go.Figure()
    fig.add_trace(go.Scatter(
        x=series.index,
        y=series,
        name="Histórico"
    ))
    fig.add_trace(go.Scatter(
        x=test.index,
        y=forecast,
        name="Evaluación pronóstico"
    ))
    fig.update_layout(title=title, template="plotly_white")
    fig.show()

In [35]:
plot_evaluation(X_data_process_df["Price"], (2,1,2), title="Materia prima X")
plot_evaluation(Y_data_process_df["Price"], (1,1,1), title="Materia prima Y")
plot_evaluation(Z_data_process_df["Price"], (2,1,2), title="Materia prima Z")

Evaluando los datos, obtenemos un MAPE < 10%, es bueno y nos permitirá tener cierta confianza en los datos.

##### Llevar a las materias primas Y y Z a la misma fecha de inicio del proyecto (basándonos en la fecha fin de registros de la materia prima X)

In [36]:
max_X = X_data_process_df.index.max()
max_Y = Y_data_process_df.index.max()
max_Z = Z_data_process_df.index.max()

gap_Y = int(np.busday_count(max_Y.date(), max_X.date()))
gap_Z = int(np.busday_count(max_Z.date(), max_X.date()))

print("Gap Y:", gap_Y)
print("Gap Z:", gap_Z)

Gap Y: 147
Gap Z: 155


In [37]:
if gap_Y > 0:
    Y_extended = pd.concat([
        Y_data_process_df["Price"],
        arima_Y.forecast(gap_Y)
    ])
else:
    Y_extended = Y_data_process_df["Price"]

if gap_Z > 0:
    Z_extended = pd.concat([
        Z_data_process_df["Price"],
        arima_Z.forecast(gap_Z)
    ])
else:
    Z_extended = Z_data_process_df["Price"]

Ahora, reentrenaremos los modelos ya alineados de materia prima Y y Z, ya que X no lo requiere:

In [38]:
arima_Y_aligned = ARIMA(
    Y_extended,
    order=(1,1,1),
    trend="t"
).fit(method="innovations_mle")

arima_Z_aligned = ARIMA(
    Z_extended,
    order=(2,1,2),
    trend="t"
).fit(method="innovations_mle")

##### Calcular días hábiles exactos a pronosticar (36 meses)

In [ ]:
start_date = X_data_process_df.index.max()
end_date = start_date + pd.DateOffset(months=36)

business_days = pd.bdate_range(start=start_date + pd.Timedelta(days=1), end=end_date)
steps_total = len(business_days)

print("Fecha inicio:", start_date)
print("Fecha fin:", end_date)
print("Días hábiles a pronosticar:", steps_total)

Fecha inicio: 2024-04-04 00:00:00
Fecha fin: 2027-04-04 00:00:00
Días hábiles a pronosticar: 781


#### Pronóstico para los 36 meses

In [40]:
forecast_X = arima_X.forecast(steps_total)
forecast_Y = arima_Y_aligned.forecast(steps_total)
forecast_Z = arima_Z_aligned.forecast(steps_total)

#### Aplicación de la segunda parte (Análisis de escenarios)

In [41]:
def monte_carlo_structural(model, steps, simulations=1000):
    simulations_matrix = []

    for _ in range(simulations):
        simulated = model.simulate(
            nsimulations=steps,
            anchor="end"
        )
        simulations_matrix.append(simulated.values)

    return np.array(simulations_matrix)

In [42]:
monte_carlo_X = monte_carlo_structural(arima_X, steps_total)
monte_carlo_Y = monte_carlo_structural(arima_Y_aligned, steps_total)
monte_carlo_Z = monte_carlo_structural(arima_Z_aligned, steps_total)

#### Construcción de equipos

In [43]:
monte_carlo_team_1 = 0.2 * monte_carlo_X + 0.8 * monte_carlo_Y
monte_carlo_team_2 = (monte_carlo_X + monte_carlo_Y + monte_carlo_Z) / 3

In [44]:
def get_metrics(monte_carlo_matrix):
    final_values = monte_carlo_matrix[:, -1]
    return {
        "Media": np.mean(final_values),
        "Desviación estandard": np.std(final_values),
        "Min": np.min(final_values),
        "Max": np.max(final_values),
        "P5": np.percentile(final_values, 5),
        "P50": np.percentile(final_values, 50),
        "P95": np.percentile(final_values, 95)
    }

metrics_2_team_1 = get_metrics(monte_carlo_team_1)
print(metrics_2_team_1)
metrics_2_team_2 = get_metrics(monte_carlo_team_2)
print(metrics_2_team_2)

{'Media': np.float64(455.4076732095588), 'Desviación estandard': np.float64(304.1583516196059), 'Min': np.float64(-539.85291606024), 'Max': np.float64(1726.7792330592886), 'P5': np.float64(-51.369265734553046), 'P50': np.float64(451.02170972289156), 'P95': np.float64(941.2628345487767)}
{'Media': np.float64(932.11404251169), 'Desviación estandard': np.float64(300.5100724384659), 'Min': np.float64(-23.61437771027887), 'Max': np.float64(1953.8961862522708), 'P5': np.float64(459.5419056739655), 'P50': np.float64(919.3496284223658), 'P95': np.float64(1448.997742420672)}


In [46]:
def plot_monte_carlo(mean_path, p5, p95, dates, title):
    fig = go.Figure()
    fig.add_trace(go.Scatter(
        x=dates,
        y=p95,
        line=dict(width=0),
        showlegend=False
    ))
    fig.add_trace(go.Scatter(
        x=dates,
        y=p5,
        fill='tonexty',
        name="Banda 90%",
        line=dict(width=0)
    ))
    fig.add_trace(go.Scatter(
        x=dates,
        y=mean_path,
        name="Escenario esperado",
        line=dict(width=2)
    ))

    fig.update_layout(
        title=dict(
            text=title,
            x=0.5,
            xanchor='center'
        ),
        template="plotly_white",
        hovermode="x unified",
        legend=dict(
            orientation="h",
            y=-0.2
        )
    )
    fig.show()

plot_monte_carlo(
    monte_carlo_team_1.mean(axis=0),
    np.percentile(monte_carlo_team_1, 5, axis=0),
    np.percentile(monte_carlo_team_1, 95, axis=0),
    business_days,
    "Diagrama de costos del Equipo 1 desde la fecha de inicio"
)

plot_monte_carlo(
    monte_carlo_team_2.mean(axis=0),
    np.percentile(monte_carlo_team_2, 5, axis=0),
    np.percentile(monte_carlo_team_2, 95, axis=0),
    business_days,
    "Diagrama de costos del Equipo 2 desde la fecha de inicio"
)

##### Conclusiones preliminares

El modelo para cada materia prima es muy bueno de acuerdo a sus métricas de error, específicamente el modelo para la materia prima Z es mucho mejor, requere mejoras ya que como vemos en la Banda de confianza al 90%, podemos observar una amplia incertidumbre mientras estamos hablando de 3 años.

De acuerdo a la metolodogía usada, es mejor que la usada en log-retornos y la incertidumbre creciente con el tiempo, al realizarse por día, las diferencias pueden ser grandes.

##### A exportar

In [47]:
X_forecast_df = pd.DataFrame({
    "Date": business_days,
    "Price": forecast_X.values,
    "Type": "Forecast"
})

X_hist_df = X_data_process_df.reset_index()[["Date", "Price"]]
X_hist_df["Type"] = "Historical"

X_full_df = pd.concat([X_hist_df, X_forecast_df], ignore_index=True)
X_full_df = X_full_df.sort_values("Date")

In [48]:
gap_dates_Y = pd.bdate_range(
    start=Y_data_process_df.index.max() + pd.Timedelta(days=1),
    periods=gap_Y
)
Y_gap_df = pd.DataFrame({
    "Date": gap_dates_Y,
    "Price": Y_extended[-gap_Y:].values if gap_Y > 0 else [],
    "Type": "Aligned"
})
Y_forecast_df = pd.DataFrame({
    "Date": business_days,
    "Price": forecast_Y.values,
    "Type": "Forecast"
})
Y_hist_df = Y_data_process_df.reset_index()[["Date", "Price"]]
Y_hist_df["Type"] = "Historical"
Y_full_df = pd.concat([Y_hist_df, Y_gap_df, Y_forecast_df], ignore_index=True)
Y_full_df = Y_full_df.sort_values("Date")

In [49]:
gap_dates_Z = pd.bdate_range(
    start=Z_data_process_df.index.max() + pd.Timedelta(days=1),
    periods=gap_Z
)
Z_gap_df = pd.DataFrame({
    "Date": gap_dates_Z,
    "Price": Z_extended[-gap_Z:].values if gap_Z > 0 else [],
    "Type": "Aligned"
})
Z_forecast_df = pd.DataFrame({
    "Date": business_days,
    "Price": forecast_Z.values,
    "Type": "Forecast"
})
Z_hist_df = Z_data_process_df.reset_index()[["Date", "Price"]]
Z_hist_df["Type"] = "Historical"
Z_full_df = pd.concat([Z_hist_df, Z_gap_df, Z_forecast_df], ignore_index=True)
Z_full_df = Z_full_df.sort_values("Date")

In [52]:
forecast_df = pd.DataFrame({
    "Date": business_days,
    "X_forecast": forecast_X.values,
    "Y_forecast": forecast_Y.values,
    "Z_forecast": forecast_Z.values
})

team_1_expected = 0.2 * forecast_X.values + 0.8 * forecast_Y.values
team_2_expected = (forecast_X.values + forecast_Y.values + forecast_Z.values) / 3
teams_expected_df = pd.DataFrame({
    "Date": business_days,
    "Team_1_expected": team_1_expected,
    "Team_2_expected": team_2_expected
})

def build_percentile_df(monte_carlo_matrix, dates):

    return pd.DataFrame({
        "Date": dates,
        "P5": np.percentile(monte_carlo_matrix, 5, axis=0),
        "P50": np.percentile(monte_carlo_matrix, 50, axis=0),
        "P95": np.percentile(monte_carlo_matrix, 95, axis=0)
    })
team_1_percentiles = build_percentile_df(monte_carlo_team_1, business_days)
team_2_percentiles = build_percentile_df(monte_carlo_team_2, business_days)

summary_df = pd.DataFrame([
    {
        "Equipo": "Equipo 1",
        **metrics_2_team_1
    },
    {
        "Equipo": "Equipo 2",
        **metrics_2_team_2
    }
])

In [54]:
X_full_df.to_csv("archivos/X_full.csv", index=False)
Y_full_df.to_csv("archivos/Y_full.csv", index=False)
Z_full_df.to_csv("archivos/Z_full.csv", index=False)

teams_expected_df.to_csv("archivos/Teams_expected.csv", index=False)

team_1_percentiles.to_csv("archivos/Team_1_percentiles.csv", index=False)
team_2_percentiles.to_csv("archivos/Team_2_percentiles.csv", index=False)

summary_df.to_csv("archivos/Teams_summary_36m.csv", index=False)

## Conclusiones

### Proceso realizado

#### Enfoque metodológico adoptado

El objetivo era estimar costos esperados y su rango de incertidumbre para planificación presupuestal a 36 meses.

Por ello:

1. Se trabajó en niveles de precios, para mantener una interpretación directa en términos monetarios.
2. Se verificó que las series de precios eran integradas de orden 1 (I(1)). Para seleccionar un modelo que se ajustara de forma preliminar a esto, ARIMA.
3. Se ajustaron modelos ARIMA(p,1,q) con drift, seleccionando hiperparámetros por AIC.
4. Se validó el desempeño fuera de muestra por 1 año en días hábiles (252 días).
5. Se proyectaron 36 meses desde el final de registros de la materia prima X, de acuerdo a los supuestos.
6. Se realizó simulación Monte Carlo estructural usando la dinámica del modelo.

    Este proceso permite:
    1. Forecast puntual (escenario esperado)
    2. Distribución probabilística (riesgo)
    3. Percentiles para toma de decisiones

#### Validación del modelo

Las métricas fuera de muestra fueron:

| Materia prima   | MAPE |
|----------|------|
| X      | 6.99%   |
| Y   | 5.67%   |
| Z    | 4.50%   |

De acuerdo a contextos:
- <10% es bueno.
- <7% permite mayor confiabilidad.

Conclusión técnica:
- El modelo reproduce adecuadamente la dinámica histórica reciente y es confiable para proyecciones estructurales.

#### Resultados – Equipo 1

Composición:

`20% X + 80% Y`

Resultados finales:
- Media ≈ 430
- P50 ≈ 422
- P95 ≈ 920
- P5 ≈ -80

Interpretación de acuerdo al contexto:
- El escenario esperado es relativamente estable.
- La banda es amplia porque el horizonte es largo (36 meses).
- La dispersión se debe principalmente a la volatilidad histórica de X.
- El percentil negativo (P5) no implica costo negativo real, sino que el supuesto estadístico (ruido normal) permite trayectorias extremas matemáticamente posibles.

Existe una variabilidad considerable en el costo del Equipo 1, con un rango probable amplio debido a alta sensibilidad a la materia prima X.

#### Resultados – Equipo 2

Composición:

`( X + Y + Z ) / 3`

Resultados finales:
- Media ≈ 914
- P50 ≈ 914
- P95 ≈ 1409
- P5 ≈ 430

Interpretación de acuerdo al contexto:
- El Equipo 2 presenta menor riesgo relativo.
- La diversificación entre tres insumos reduce extremos.
- La distribución es más simétrica y económicamente coherente.
- No aparecen valores negativos en percentiles relevantes.

El Equipo 2 es estructuralmente más estable ante shocks individuales, debido a diversificación de materias primas.

#### ¿Por qué la banda es amplia?

Hay tres razones estructurales:
1. Horizonte largo (36 meses).
2. Series I(1): la varianza crece con el tiempo (metodología tomada).
3. Volatilidad histórica significativa en materias primas.

Pese a que parte de esto se refleja por el riesgo real del mercado, el modelo puede mejorar para brindar aún más precisión y confianza.

#### Respecto a los beneficios esperados

##### Precisión en estimación de costos

- El modelo logra un error histórico bajo (<7%).
- Se obtiene un valor esperado robusto.
- Se cuantifica incertidumbre.
- Se identifican riesgos de desviación presupuestal.

Esto cumple el primer beneficio esperado.

##### Optimización de recursos

El análisis permite:

- Evaluar qué equipo es más sensible a volatilidad.
- Determinar probabilidad de sobrecostos.
- Diseñar estrategias de compra anticipada.
- Evaluar cobertura financiera.

Esto cumple el segundo beneficio esperado.